### PR TITLE
New View: X Family Tree

### PR DESCRIPTION
**Pull request by Clarke-11007**
This new Tree App will show you which are your X-Chromosome ancestors (potentially), and build a family tree based on that specific inheritance pattern.

There are two modes:  Probability mode which shows the % inherited from each if at every point 50% of each parent's X-Chromosome was passed on during a recombination (from Mother to child).

Random Simulation mode assigns random recombination points to the chromosomes and simulates how the various pieces get passed down, which you can Randomize and repeat over and over again for fun!

### CODE REVIEW

**Comment by MichalVasut** [comment]
Is there any preview where can I see what it does? (with some testing data, I don't have DNA tests)

I'm roughly guessing from the code, but I'm not able to visualize it.

**Comment by Clarke-11007** [comment]
Great question Michal - I'm bad about not thinking about that.

Here is a link to my dev site (on the apps server)

https://apps.wikitree.com/apps/clarke11007/WTdynamicTree/#name=Clarke-11007&view=xtree

**Comment by MichalVasut** [comment]
Cool, thanks - btw, your video was very useful 👍

https://youtu.be/ppO7tzxuik8

File: views/xTree/XTreeView.js [hunk 1]
```diff
@@ -0,0 +1,1323 @@
...
+                if (person.DataStatus.BirthDate == "guess") {
+                    html += " about ";
+                }
+                if (person.DataStatus.BirthDate == "before") {
+                    html += " before ";
+                }
+                if (person.DataStatus.BirthDate == "after") {
+                    html += " after ";
+                }
+                html += ` ${this.formatDate(person.BirthDate)}`;
```
Comment (MichalVasut) [inline views/xTree/XTreeView.js:1274]: just simplification

```js
html += ["guess", "before", "after"].includes(person.DataStatus.BirthDate) ? ` ${person.DataStatus.BirthDate} ` : ""
```

File: views/xTree/XTreeView.js [hunk 2]
```diff
@@ -0,0 +1,1323 @@
...
+                if (person.Nicknames) {
+                    html += `"${person.Nicknames}" `;
+                }
+                if (person.LastNameCurrent != person.LastNameAtBirth) {
+                    html += ` (${person.LastNameAtBirth}) `;
+                }
+                html += `${person.LastNameCurrent}`;
+                if (person.Suffix) {
+                    html += ` ${person.Suffix}`;
+                }
```
Comment (MichalVasut) [inline views/xTree/XTreeView.js:1238]: btw, you can use your function `displayPersonName` (also in `else` branch of this `if` statement) - it seems very similar

File: views/xTree/XTreeView.js [hunk 3]
```diff
@@ -0,0 +1,1323 @@
...
+
+        // Display our "info" panel with a description of this view.
+        wtViewRegistry.showInfoPanel();
+
+        // Setup the Button Bar --> Initial version will use mostly text links, but should be replaced with icons - ideally images that have a highlighted / unhighlighted version, where appropriate
+        let btnBarHTML =
+            '<table border=0 style="background-color: #f8a51d80;" width="100%"><tr>' +
+            '<td width="30%">' +
+            "&nbsp;" +
+            "<input type=radio name=appMode checked value=Probability onclick='XTreeView.changeMode(0);'>Probability Tree&nbsp;&nbsp; <input type=radio name=appMode value=Random onclick='XTreeView.changeMode(1);'>Random Simulation&nbsp;&nbsp; "+
```
Comment (MichalVasut) [inline views/xTree/XTreeView.js:369]: better use `<label>`  element for inputs ([docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/label)), it'll improve usability of the input ... click on the label triggers the input.

File: views/xTree/XTreeView.js [hunk 4]
```diff
@@ -0,0 +1,1323 @@
...
+        Though visually pleasing, this EXACT result is highly unlikely.  <br/><br/>
+        However, it should give an idea of which ancestors are more likely than others to have passed on some of their X-Chromosome,<br/>
+        and make it obvious how random DNA inheritance really is.<br/><br/>
+        Comparing your X-chromosome DNA with other relatives' DNA is the only way to confirm which ancestors truly passed on theirs to you.<br/>`;
+
+        XTreeView.forMore = `<br/><hr/>For more details about the X-Chromosome inheritance pattern, check out the <a target=_blank href="https://youtu.be/ppO7tzxuik8?t=64">intro to this video</a>.<br/>
+        You can display your own X-Chromosome ancestors in the <a target=_blank href="#view=fanchart">Fan Chart</a> &rarr; open the Settings, choose Highlights, then Highlight cells by <I>X-chromosome inheritance</I>.<br/>
+        To investigate further and find other relatives that could share some of your ancestors X-Chromosomes, try the <a target=_blank href="https://apps.wikitree.com/apps/clarke11007/Xfriends.php">X-Friends app</a>.`;
```
Comment (MichalVasut) [inline views/xTree/XTreeView.js:459]: here is just an idea (when I see those large paragraphs of plain text) - there is posibility to store it in text file and load it "on the fly"

```js
await (await fetch("tree.js")).text()
```

(you can try it eg at wikitree.com in js console 😉)

Comment (Clarke-11007) [reply views/xTree/XTreeView.js:459]: Thanks for all those suggestions Michal!  Some of the code was hanging around from another view that I "stole" this setup from ... so I can't claim it was all original - but - many of your suggestions WERE mine.  I appreciate you helping me make it more efficient.  The SVG using CSS ... I kind of knew that, but hadn't thought it through to the logical conclusion - makes a lot of sense.

**Comment by MichalVasut** [comment]
@Clarke-11007 I've noticed, that Settings gear button doesn't work, showing 

> Uncaught TypeError: XTreeView.toggleSettings is not a function at HTMLAnchorElement.onclick

in JS console, after clicking. After brief code inspection, the function `XTreeView.toggleSettings` is really missing. Maybe it woud be better to temporarily hide the button (e.g. `display: None`), to not confuse users.

**Comment by Clarke-11007** [comment]
Thanks for finding that Michal!  I'll do a quick PR update later today

 -Greg

**Pull request merged**